# Supervised Learning Model Evaluation Lab

Complete the exercises below to solidify your knowledge and understanding of supervised learning model evaluation.

In [1]:
import pandas as pd

## Regression Model Evaluation

In [2]:
from sklearn.datasets import load_boston

data = load_boston()

X = pd.DataFrame(data["data"], columns=data["feature_names"])
y = pd.DataFrame(data["target"], columns=['MEDV'])

data = pd.concat([X, y], axis=1)

## 1. Split this data set into training (80%) and testing (20%) sets.

The `MEDV` field represents the median value of owner-occupied homes (in $1000's) and is the target variable that we will want to predict.

In [3]:
from sklearn.model_selection import train_test_split

In [4]:
X_train_bos, X_test_bos, y_train_bos, y_test_bos = train_test_split(X, y, test_size=0.2, random_state=5)

## 2. Train a `LinearRegression` model on this data set and generate predictions on both the training and the testing set.

In [5]:
from sklearn.linear_model import LinearRegression

In [6]:
boston_lr = LinearRegression()

In [7]:
boston_lr.fit(X_train_bos,y_train_bos)

LinearRegression()

In [8]:
pred_test_bos = boston_lr.predict(X_test_bos)

In [9]:
pred_train_bos = boston_lr.predict(X_train_bos)

## 3. Calculate and print R-squared for both the training and the testing set.

In [10]:
from sklearn.metrics import r2_score

In [11]:
r2_score(y_train_bos, pred_train_bos)

0.738339392059052

In [12]:
r2_score(y_test_bos, pred_test_bos)

0.7334492147453092

## 4. Calculate and print mean squared error for both the training and the testing set.

In [13]:
from sklearn.metrics import mean_squared_error

In [14]:
mean_squared_error(y_train_bos, pred_train_bos)

22.477090408387628

In [15]:
mean_squared_error(y_test_bos, pred_test_bos)

20.869292183770682

## 5. Calculate and print mean absolute error for both the training and the testing set.

In [16]:
from sklearn.metrics import mean_absolute_error

In [17]:
mean_absolute_error(y_train_bos, pred_train_bos)

3.3500095196484505

In [18]:
mean_absolute_error(y_test_bos, pred_test_bos)

3.21327049584237

## Classification Model Evaluation

In [19]:
from sklearn.datasets import load_iris

data = load_iris()

X = pd.DataFrame(data["data"], columns=data["feature_names"])
y = pd.DataFrame(data["target"], columns=["class"])

data = pd.concat([X, y], axis=1)

## 6. Split this data set into training (80%) and testing (20%) sets.

The `class` field represents the type of flower and is the target variable that we will want to predict.

In [20]:
X_train_iris, X_test_iris, y_train_iris, y_test_iris = train_test_split(X, y, test_size=0.2, random_state=5)

## 7. Train a `LogisticRegression` model on this data set and generate predictions on both the training and the testing set.

In [21]:
from sklearn.linear_model import LogisticRegression

In [22]:
log = LogisticRegression(max_iter = 1000)

In [23]:
y_train_iris = y_train_iris['class'] #for removing Warning about array type

In [24]:
log.fit(X_train_iris,y_train_iris)

LogisticRegression(max_iter=1000)

In [25]:
pred_test_iris = log.predict(X_test_iris)

In [26]:
pred_train_iris = log.predict(X_train_iris)

## 8. Calculate and print the accuracy score for both the training and the testing set.

In [27]:
from sklearn.metrics import accuracy_score

In [28]:
accuracy_score(y_train_iris,pred_train_iris )

0.9833333333333333

In [29]:
accuracy_score(y_test_iris,pred_test_iris)

0.9666666666666667

## 9. Calculate and print the balanced accuracy score for both the training and the testing set.

In [30]:
from sklearn.metrics import balanced_accuracy_score

In [31]:
balanced_accuracy_score(y_train_iris,pred_train_iris)

0.982905982905983

In [32]:
balanced_accuracy_score(y_test_iris,pred_test_iris)

0.9696969696969697

## 10. Calculate and print the precision score for both the training and the testing set.

In [33]:
from sklearn.metrics import precision_score

In [34]:
precision_score(y_train_iris,pred_train_iris, average = None)

array([1.        , 1.        , 0.95121951])

In [35]:
precision_score(y_test_iris,pred_test_iris, average = None)

array([1.        , 1.        , 0.91666667])

## 11. Calculate and print the recall score for both the training and the testing set.

In [36]:
from sklearn.metrics import recall_score

In [37]:
recall_score(y_train_iris,pred_train_iris, average = None)

array([1.        , 0.94871795, 1.        ])

In [38]:
recall_score(y_test_iris,pred_test_iris, average = None)

array([1.        , 0.90909091, 1.        ])

## 12. Calculate and print the F1 score for both the training and the testing set.

In [39]:
from sklearn.metrics import f1_score

In [40]:
f1_score(y_train_iris,pred_train_iris, average = None)

array([1.        , 0.97368421, 0.975     ])

In [41]:
f1_score(y_test_iris,pred_test_iris, average = None)

array([1.        , 0.95238095, 0.95652174])

## 13. Generate confusion matrices for both the training and the testing set.

In [42]:
from sklearn.metrics import confusion_matrix

In [43]:
confusion_matrix(y_train_iris,pred_train_iris)

array([[42,  0,  0],
       [ 0, 37,  2],
       [ 0,  0, 39]])

In [44]:
confusion_matrix(y_test_iris,pred_test_iris)

array([[ 8,  0,  0],
       [ 0, 10,  1],
       [ 0,  0, 11]])

## Bonus: For each of the data sets in this lab, try training with some of the other models you have learned about, recalculate the evaluation metrics, and compare to determine which models perform best on each data set.

### Boston dataset

In [45]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

In [46]:
rfr = RandomForestRegressor()

In [47]:
X_train_bos.shape

(404, 13)

In [48]:
parameter_space = {'n_estimators': [100, 300, 1000],
                   'max_features': ['sqrt', 0.6, None],
                   'max_depth': [None, 10, 30, 100],
                   'min_samples_leaf': [1, 3, 10]}

grid_search = GridSearchCV(rfr,
                           param_grid=parameter_space,
                           verbose=1,
                           n_jobs=-1,
                           cv=5)

grid_search.fit(X_train_bos, y_train_bos)

Fitting 5 folds for each of 108 candidates, totalling 540 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  26 tasks      | elapsed:    3.0s
[Parallel(n_jobs=-1)]: Done 176 tasks      | elapsed:   16.1s
[Parallel(n_jobs=-1)]: Done 426 tasks      | elapsed:   41.2s
[Parallel(n_jobs=-1)]: Done 540 out of 540 | elapsed:   54.2s finished
/home/sergio/anaconda3/lib/python3.8/site-packages/sklearn/model_selection/_search.py:765: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  self.best_estimator_.fit(X, y, **fit_params)


GridSearchCV(cv=5, estimator=RandomForestRegressor(), n_jobs=-1,
             param_grid={'max_depth': [None, 10, 30, 100],
                         'max_features': ['sqrt', 0.6, None],
                         'min_samples_leaf': [1, 3, 10],
                         'n_estimators': [100, 300, 1000]},
             verbose=1)

In [49]:
grid_search.best_score_

0.8762561193974682

In [50]:
best_rf = grid_search.best_estimator_

In [51]:
best_rf.score(X_test_bos, y_test_bos)

0.8906723918828238

With Random Forest R2 = 0.8906723918828238 vs R2 = 0.7334492147453092   for Linear Regression.

This random forest is better model.

Other metrics:

In [52]:
pred_boston_test = best_rf.predict(X_test_bos)

In [53]:
mean_squared_error(y_test_bos, pred_boston_test)

8.559681395686221

In [54]:
mean_absolute_error(y_test_bos, pred_boston_test)

1.9948999999999903

### Iris dataset

In [55]:
from sklearn.ensemble import RandomForestClassifier

In [56]:
rfc = RandomForestClassifier()

In [57]:
X_train_iris.shape

(120, 4)

In [58]:
parameter_space = {'n_estimators': [100, 300, 1000],
                   'max_features': ['sqrt', None],
                   'max_depth': [None, 10, 30, 100],
                   'min_samples_leaf': [1, 3, 10]}

grid_search2 = GridSearchCV(rfc,
                           param_grid=parameter_space,
                           verbose=1,
                           n_jobs=-1,
                           cv=5)

grid_search2.fit(X_train_iris, y_train_iris)

Fitting 5 folds for each of 72 candidates, totalling 360 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  29 tasks      | elapsed:    2.0s
[Parallel(n_jobs=-1)]: Done 212 tasks      | elapsed:   17.0s
[Parallel(n_jobs=-1)]: Done 360 out of 360 | elapsed:   29.5s finished


GridSearchCV(cv=5, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'max_depth': [None, 10, 30, 100],
                         'max_features': ['sqrt', None],
                         'min_samples_leaf': [1, 3, 10],
                         'n_estimators': [100, 300, 1000]},
             verbose=1)

In [59]:
best_rf2 = grid_search2.best_estimator_

In [60]:
best_rf2.score(X_test_iris, y_test_iris)

0.9

With Random Forest accuracy = 0.9 vs R2 = 0.9666666666666667   for 
Logistic Regression.

Logistic Regression is better model.

Ohter metrics:

In [61]:
pred_iris_test = best_rf2.predict(X_test_iris)

In [62]:
pred_test_iris

array([1, 2, 2, 0, 2, 1, 0, 2, 0, 1, 1, 2, 2, 2, 0, 0, 2, 2, 0, 0, 1, 2,
       0, 1, 1, 2, 1, 1, 1, 2])

In [64]:
balanced_accuracy_score(y_test_iris,pred_test_iris)

0.9696969696969697

In [69]:
precision_score(y_test_iris,pred_iris_test, average = None)

array([1.        , 0.83333333, 0.9       ])

In [70]:
recall_score(y_test_iris,pred_iris_test, average = None)

array([1.        , 0.90909091, 0.81818182])

In [71]:
f1_score(y_test_iris,pred_iris_test, average = None)

array([1.        , 0.86956522, 0.85714286])

In [72]:
confusion_matrix(y_test_iris,pred_iris_test)

array([[ 8,  0,  0],
       [ 0, 10,  1],
       [ 0,  2,  9]])